packages/plugins/src/menu/Export104.ts

```diff
@@ -0,0 +1,90 @@
+import { LitElement, property } from 'lit-element';
+import { stringify } from 'csv-stringify/browser/esm/sync';
+import { newLogEvent } from '@openscd/core/foundation/deprecated/history.js';
+
+import { extractAllSignal104Data, Signal104 } from './export104/foundation.js';
+import { get } from 'lit-translate';
+
+
+
+export default class Export104 extends LitElement {
+  @property({ attribute: false }) doc!: XMLDocument;
+  @property() docName!: string;
+
+  private readonly csvHeaders = [
+    'Id',
+    'Name',
+    'Signal Number',
+    'mIOA',
+    'cIOA'
+  ];
+
+  async run(): Promise<void> {
+    const { signals, errors } = extractAllSignal104Data(this.doc);
+
+    errors.forEach((error) => this.logWarning(error));
+
+    if (signals.length === 0) {
+      this.dispatchEvent(newLogEvent({
+        kind: 'info',
+        title: get('protocol104.export.noSignalsFound'),
+      }));
+      return;
+    }
+
+    const csvLines = this.generateCsvLines(signals);
+
+    const csvContent = stringify(csvLines, {
+      header: true,
+      columns: this.csvHeaders,
+    });
+    const csvBlob = new Blob([csvContent], {
+      type: 'text/csv',
+    });
+
+    this.downloadCsv(csvBlob);
+  }
+
+  private logWarning(errorMessage: string): void {
+    this.dispatchEvent(newLogEvent({
+      kind: 'warning',
+      title: get('protocol104.export.invalidSignalWarning'),
+      message: errorMessage,
+    }));
+  }
+
+  private generateCsvLines(allSignal104Data: Signal104[]): string[][] {
+    const lines: string[][] = [];
+
+    for(const signal104Data of allSignal104Data) {
+      const line = [
+        '',
+        signal104Data.name ?? '',
+        signal104Data.signalNumber ?? '',
+      ];
+
+      if (signal104Data.isMonitorSignal) {
+        line.push(signal104Data.ioa ?? '', '');
+      } else {
+        line.push('', signal104Data.ioa ?? '');
+      }
+
+      lines.push(line);
+    }
+
+    return lines;
+  }
+
+  private downloadCsv(csvBlob: Blob): void {
+    const a = document.createElement('a');
+    a.download = this.docName + '-104-signals.csv';
+    a.href = URL.createObjectURL(csvBlob);
+    a.dataset.downloadurl = ['text/csv', a.download, a.href].join(':');
+    a.style.display = 'none';
+
+    document.body.appendChild(a);
+    a.click();
+    document.body.removeChild(a);
+    URL.revokeObjectURL(a.href);
```

Comment:
This is a nice hack. I assume there are not other ways to do this

Reply:
I pretty much copied the code from here https://github.com/com-pas/compas-open-scd/blob/main/packages/compas-open-scd/src/menu/ExportIEDParams.ts#L296

After googling for a while the hidden `<a>` element seems to be the only way, we might want to look for a way to make this function reuseable, because at the moment it has been copied at least three times between different plugins.